MAINT, TST: test_public_api.py can fail with NumPy main, via CuPy

On latest NumPy `main` alongside SciPy `main`, with `python dev.py test -j 32` (no array API stuff activated), I get an error that traverses through CuPy in `test_public_api.py`. I think my argument here is that I should be able to have pretty much any old version of CuPy installed that I want if I'm not opting in to the array API stuff when testing, at least for now.

Traceback below the fold

<details>

```
================================================ FAILURES =================================================
______________________________________ test_all_modules_are_expected ______________________________________
[gw0] linux -- Python 3.11.2 /home/treddy/python_venvs/py_311_scipy_dev/bin/python
scipy/_lib/tests/test_public_api.py:222: in test_all_modules_are_expected
    for _, modname, ispkg in pkgutil.walk_packages(path=scipy.__path__,
        _          = FileFinder('/home/treddy/github_projects/scipy/build-install/lib/python3.11/site-packages/scipy/_lib/array_api_compat')
        ispkg      = True
        modname    = 'scipy._lib.array_api_compat.cupy'
        modnames   = []
../../../../../spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.3.0/python-3.11.2-4qncg2nqev5evxfmamde3e6rnb34b4ls/lib/python3.11/pkgutil.py:107: in walk_packages
    yield from walk_packages(path, info.name+'.', onerror)
        info       = ModuleInfo(module_finder=FileFinder('/home/treddy/github_projects/scipy/build-install/lib/python3.11/site-packages/scipy'), name='scipy._lib', ispkg=True)
        onerror    = None
        path       = ['/home/treddy/github_projects/scipy/build-install/lib/python3.11/site-packages/scipy/_lib']
        prefix     = 'scipy.'
        seen       = <function walk_packages.<locals>.seen at 0x7fd2a8311800>
../../../../../spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.3.0/python-3.11.2-4qncg2nqev5evxfmamde3e6rnb34b4ls/lib/python3.11/pkgutil.py:107: in walk_packages
    yield from walk_packages(path, info.name+'.', onerror)
        info       = ModuleInfo(module_finder=FileFinder('/home/treddy/github_projects/scipy/build-install/lib/python3.11/site-packages/scipy/_lib'), name='scipy._lib.array_api_compat', ispkg=True)
        onerror    = None
        path       = ['/home/treddy/github_projects/scipy/build-install/lib/python3.11/site-packages/scipy/_lib/array_api_compat']
        prefix     = 'scipy._lib.'
        seen       = <function walk_packages.<locals>.seen at 0x7fd2a8312ac0>
../../../../../spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.3.0/python-3.11.2-4qncg2nqev5evxfmamde3e6rnb34b4ls/lib/python3.11/pkgutil.py:92: in walk_packages
    __import__(info.name)
        info       = ModuleInfo(module_finder=FileFinder('/home/treddy/github_projects/scipy/build-install/lib/python3.11/site-packages/scipy/_lib/array_api_compat'), name='scipy._lib.array_api_compat.cupy', ispkg=True)
        onerror    = None
        path       = ['/home/treddy/github_projects/scipy/build-install/lib/python3.11/site-packages/scipy/_lib/array_api_compat/common']
        prefix     = 'scipy._lib.array_api_compat.'
        seen       = <function walk_packages.<locals>.seen at 0x7fd2a8312c00>
scipy/_lib/array_api_compat/cupy/__init__.py:1: in <module>
    from cupy import *
        __builtins__ = <builtins>
        __cached__ = '/home/treddy/github_projects/scipy/build-install/lib/python3.11/site-packages/scipy/_lib/array_api_compat/cupy/__pycache__/__init__.cpython-311.pyc'
        __doc__    = None
        __file__   = '/home/treddy/github_projects/scipy/build-install/lib/python3.11/site-packages/scipy/_lib/array_api_compat/cupy/__init__.py'
        __loader__ = <_frozen_importlib_external.SourceFileLoader object at 0x7fd2a83326d0>
        __name__   = 'scipy._lib.array_api_compat.cupy'
        __package__ = 'scipy._lib.array_api_compat.cupy'
        __path__   = ['/home/treddy/github_projects/scipy/build-install/lib/python3.11/site-packages/scipy/_lib/array_api_compat/cupy']
        __spec__   = ModuleSpec(name='scipy._lib.array_api_compat.cupy', loader=<_frozen_importlib_external.SourceFileLoader object at 0x7f...ons=['/home/treddy/github_projects/scipy/build-install/lib/python3.11/site-packages/scipy/_lib/array_api_compat/cupy'])
/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/__init__.py:17: in <module>
    from cupy import _core  # NOQA
        __builtins__ = <builtins>
        __cached__ = '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/__pycache__/__init__.cpython-311.pyc'
        __doc__    = None
        __file__   = '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/__init__.py'
        __loader__ = <_frozen_importlib_external.SourceFileLoader object at 0x7fd2a8332bd0>
        __name__   = 'cupy'
        __package__ = 'cupy'
        __path__   = ['/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy']
        __spec__   = ModuleSpec(name='cupy', loader=<_frozen_importlib_external.SourceFileLoader object at 0x7fd2a8332bd0>, origin='/home/t...nit__.py', submodule_search_locations=['/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy'])
        _environment = <module 'cupy._environment' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/_environment.py'>
        _functools = <module 'functools' from '/home/treddy/github_projects/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.3.0/python-3.11.2-4qncg2nqev5evxfmamde3e6rnb34b4ls/lib/python3.11/functools.py'>
        _numpy     = <module 'numpy' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/numpy/__init__.py'>
        _sys       = <module 'sys' (built-in)>
        _util      = <module 'cupy._util' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/_util.cpython-311-x86_64-linux-gnu.so'>
        _version   = <module 'cupy._version' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/_version.py'>
        cuda       = <module 'cupy.cuda' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/cuda/__init__.py'>
/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/_core/__init__.py:3: in <module>
    from cupy._core import core  # NOQA
        __builtins__ = <builtins>
        __cached__ = '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/_core/__pycache__/__init__.cpython-311.pyc'
        __doc__    = None
        __file__   = '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/_core/__init__.py'
        __loader__ = <_frozen_importlib_external.SourceFileLoader object at 0x7fd2a833ed10>
        __name__   = 'cupy._core'
        __package__ = 'cupy._core'
        __path__   = ['/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/_core']
        __spec__   = ModuleSpec(name='cupy._core', loader=<_frozen_importlib_external.SourceFileLoader object at 0x7fd2a833ed10>, origin='/...py', submodule_search_locations=['/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/_core'])
        _carray    = <module 'cupy._core._carray' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/_core/_carray.cpython-311-x86_64-linux-gnu.so'>
        _dtype     = <module 'cupy._core._dtype' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/_core/_dtype.cpython-311-x86_64-linux-gnu.so'>
        internal   = <module 'cupy._core.internal' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/_core/internal.cpython-311-x86_64-linux-gnu.so'>
        syncdetect = <module 'cupy._core.syncdetect' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/_core/syncdetect.py'>
cupy/_core/core.pyx:1: in init cupy._core.core
    ???
cupy/_core/_scalar.pyx:64: in init cupy._core._scalar
    ???
/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/numpy/__init__.py:393: in __getattr__
    raise AttributeError(
E   AttributeError: `np.float_` was removed in the NumPy 2.0 release. Use `np.float64` instead.
        attr       = 'float_'
        warnings   = <module 'warnings' from '/home/treddy/github_projects/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.3.0/python-3.11.2-4qncg2nqev5evxfmamde3e6rnb34b4ls/lib/python3.11/warnings.py'>
```

</details>

So, yeah, maybe something to be patched in CuPy eventually, but for now I should be able to run the testsuite just fine in this scenario I'd say.

--- a/scipy/_lib/tests/test_public_api.py
+++ b/scipy/_lib/tests/test_public_api.py
@@ -13,6 +13,8 @@ import pytest
 
 import scipy
 
+from scipy.conftest import xp_available_backends
+
 
 def test_dir_testing():
     """Assert that output of dir has only one "testing/tester"
@@ -212,16 +214,33 @@ SKIP_LIST = [
 ]
 
 
+# XXX: this test does more than it says on the tin - in using `pkgutil.walk_packages`,
+# it will raise if it encounters any exceptions which are not of type `ImportError` or
+# handled by `ignore_errors` while attempting to import each discovered package.
+# For now, `ignore_errors` only ignores what is necessary, but this could be expanded -
+# for example, to all errors from private modules or git subpackages - if desired.
 def test_all_modules_are_expected():
     """
     Test that we don't add anything that looks like a new public module by
     accident.  Check is based on filenames.
     """
 
+    def ignore_errors(name):
+        # if versions of other array libraries are installed which are incompatible
+        # with the installed NumPy version, there can be errors on importing
+        # `array_api_compat`. This should only raise if SciPy is configured with
+        # that library as an available backend.
+        for backend, dir_name in {'cupy': 'cupy', 'pytorch': 'torch'}.items():
+            path = f'array_api_compat.{dir_name}'
+            if path in name and backend not in xp_available_backends:
+                return
+        raise
+
     modnames = []
-    for _, modname, ispkg in pkgutil.walk_packages(path=scipy.__path__,
-                                                   prefix=scipy.__name__ + '.',
-                                                   onerror=None):
+
+    for _, modname, _ in pkgutil.walk_packages(path=scipy.__path__,
+                                               prefix=scipy.__name__ + '.',
+                                               onerror=ignore_errors):
         if is_unexpected(modname) and modname not in SKIP_LIST:
             # We have a name that is new.  If that's on purpose, add it to
             # PUBLIC_MODULES.  We don't expect to have to add anything to

